packages/react-pages/src/node/dynamic-modules/PagesData.ts

```diff
@@ -0,0 +1,289 @@
+import deepEqual from 'fast-deep-equal'
+import { File } from './PageStrategy'
+import type { ScheduleUpdate } from './UpdateBuffer'
+
+export class PagesDataKeeper {
+  private pages: PagesDataInternal = {}
+
+  /**
+   * If the page page does not come from local file,
+   * then we create this api to track the source of page data.
+   * When a file is unlinked,
+   * the data associated with it will automatically get deleted.
+   */
+  createAPIForSourceFile(
+    sourceFile: File,
+    scheduleUpdate: ScheduleUpdate
+  ): HanlderAPI {
+    return {
+      getRuntimeData: (
+        pageId: string
+      ): {
+        [key: string]: string
+      } =>
+        this.createMutableProxy(pageId, 'runtime', sourceFile, scheduleUpdate),
+      getStaticData: (
+        pageId: string
+      ): {
+        [key: string]: any
+      } =>
+        this.createMutableProxy(pageId, 'static', sourceFile, scheduleUpdate),
+    }
+  }
+
+  /**
+   * If the page page does not come from local file (.e.g from remote database),
+   * then we can't track the source of page data.
+   */
+  createAPIForCustomSource(scheduleUpdate: ScheduleUpdate): HanlderAPI {
+    return {
+      getRuntimeData: (
+        pageId: string
+      ): {
+        [key: string]: string
+      } => this.createMutableProxy(pageId, 'runtime', null, scheduleUpdate),
+      getStaticData: (
+        pageId: string
+      ): {
+        [key: string]: any
+      } => this.createMutableProxy(pageId, 'static', null, scheduleUpdate),
+    }
+  }
+
+  private createMutableProxy(
+    pageId: string,
+    dataType: 'runtime' | 'static',
+    sourceFile: File | null,
+    scheduleUpdate: ScheduleUpdate
+  ): {
+    [key: string]: any
+  } {
+    const _this = this
+    const dataObj = this.getDataObj(pageId, dataType)
+    return new Proxy(dataObj, {
+      set(target, prop: string, value) {
+        if (value === undefined) {
+          // undefined means 'delete'
+          const existValue = dataObj[prop]
+          if (existValue) {
+            existValue.unBind(scheduleUpdate)
+          }
+          return true
+        }
+        new Association(
```

Comment:
As a general rule, constructors shouldn't have side effects.

Reply:
I also had doubt when writing this line. But I really want to have a `Association` class to represent the binding between file and pageData, and let it handle all the related works. It produces a clearer mind model.